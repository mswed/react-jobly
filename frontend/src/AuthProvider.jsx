import React, { useState, useEffect } from 'react';
import JoblyApi from './api';

// We create the context
const AuthContext = React.createContext();

// We set up a component to manage the context
const AuthProvider = ({ children }) => {
  // We initialize the token with the value from localStorage
  const [token, setToken] = useState(localStorage.getItem('jobly-token'));

  // Whenever the token changes (login/logout) we need to update
  // both the api module and localStorage
  useEffect(() => {
    JoblyApi.token = token;
    localStorage.setItem('jobly-token', token || '');
  }, [token]);

  const login = async (username, password) => {
    try {
      const token = await JoblyApi.login(username, password);
      setToken(token);
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const register = async (username, password, firstName, lastName, email) => {
    try {
      const token = await JoblyApi.register(username, password, firstName, lastName, email);
      setToken(token);
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  // Create the value object
  const value = { token, register };

  // Wrap everything in the context
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export { AuthContext, AuthProvider };
