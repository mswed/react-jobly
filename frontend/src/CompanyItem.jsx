const CompanyItem = ({ name, description, logo }) => {
  return (
    <div>
      <h4>{name}</h4>
      <p>{description}</p>
      <img src={logo} alt={`Logo for ${name}`} />
    </div>
  );
};

export default CompanyItem;
