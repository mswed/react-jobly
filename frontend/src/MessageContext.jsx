import React, { useState } from 'react';

// We create the context
const MessageContext = React.createContext();

// We set up a component to manage the context
const MessagesProvider = ({ children }) => {
  const [message, setMessage] = useState(null);

  const showMessage = (text, variant = 'success', duration = 5000) => {
    setMessage({ text, variant, duration });
    // Auto clear
    if (duration) {
      setTimeout(() => {
        clearMessage();
      }, duration);
    }
  };

  const clearMessage = () => {
    setMessage(null);
  };

  return <MessageContext.Provider value={{ message, showMessage, clearMessage }}>{children}</MessageContext.Provider>;
};

export { MessageContext, MessagesProvider };
