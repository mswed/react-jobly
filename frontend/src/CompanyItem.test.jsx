import { it, expect, vi, describe, beforeEach } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import CompanyItem from './CompanyItem';
import { MemoryRouter } from 'react-router-dom';
import { MessageContext } from './MessageContext';
import { AuthContext } from './AuthProvider';

// Mock authentication
const authValue = {
  token: 'fake-token',
  currentUser: 'testuser',
  logout: vi.fn(),
  applyToJob: vi.fn(),
  alreadyApplied: vi.fn().mockImplementation((jobId) => {
    // Mock that job with ID 1 has been applied to
    return jobId === 1;
  }),
};

// Mock the api
vi.mock('./api', () => ({
  default: {
    getCompany: vi.fn().mockResolvedValue({
      handle: 'test-company',
      name: 'Fancy Company',
      description: 'Just a test company',
      numEmployees: 10,
      logoUrl: 'http://www.test.com',
      jobs: [
        { id: 1, title: 'Test Job', salary: 100, equity: 0.15 },
        { id: 2, title: 'Another job', salary: 60000, equity: 0.3 },
      ],
    }),
  },
}));

// Mock job functions
vi.mock('.Job.alreadyApplied', () => vi.fn().mockResolvedValue(False));

describe('CompanyItem', () => {
  // Reset all mocks before each test
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders without crashing', async () => {
    render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: vi.fn() }}>
          <AuthContext.Provider value={authValue}>
            <CompanyItem name={'Sample Company'} description={'Sample Company Description'} logo={'https://www.logo.com'} />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );
  });

  it('matches snapshot', async () => {
    const { asFragment } = render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: vi.fn() }}>
          <AuthContext.Provider value={authValue}>
            <CompanyItem name={'Sample Company'} description={'Sample Company Description'} logo={'https://www.logo.com'} />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays expected content', async () => {
    render(
      <MemoryRouter>
        <MessageContext.Provider value={{ showMessage: vi.fn() }}>
          <AuthContext.Provider value={authValue}>
            <CompanyItem name={'Sample Company'} description={'Sample Company Description'} logo={'https://www.logo.com'} />
          </AuthContext.Provider>
        </MessageContext.Provider>
      </MemoryRouter>
    );
    // Wait for the companies to load
    await waitFor(() => {
      expect(screen.getByText('Sample Company')).toBeInTheDocument();
      expect(screen.getByText('Sample Company Description')).toBeInTheDocument();
      expect(screen.getByAltText('Logo for Sample Company')).toBeInTheDocument();
    });
  });
});
