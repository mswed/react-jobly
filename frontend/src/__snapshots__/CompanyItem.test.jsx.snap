// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CompanyItem > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="shadow-lg mb-3 card"
  >
    <div
      class="g-0 row"
    >
      <div
        class="col-md-8"
      >
        <div
          class="card-body"
        >
          <div
            class="card-title h5"
          >
            Sample Company
          </div>
          <p
            class="card-text"
          >
            Sample Company Description
          </p>
        </div>
      </div>
      <div
        class="col-md-4"
      >
        <img
          alt="Logo for Sample Company"
          class="card-img"
          src="https://www.logo.com"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
